### PR TITLE
i#5843 scheduler: Add replay of the as-traced schedule

### PR DESCRIPTION
**Pull request by derekbruening**
Adds replay of the as-traced schedule using the cpu_schedule.zip file written by raw2trace.  That file is converted into the record+replay format for dynamic schedules and the existing replay code leveraged.

This implements as-traced cpu-oriented parallel iteration in the scheduler which is part of #5694.  Adding analysis tool support for this will be done separately.

Updates all replays (whether as-traced or as-previously) to consider DEPENDENCY_TIMESTAMPS to indicate whether to have one output wait if it gets ahead of another's timestamp.

Removes STATUS_IDLE since we always use STATUS_WAIT.  We're between releases so we can delete from the public enum for this still-under-development component without worrying about binary compatibility.

Adds a missing check for output count mismatches on as-previously replay.

Adds unit tests of synthetic and from-file replay.

Issue: #5843, #5694

### CODE REVIEW

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 1]
```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
...
+    // designed for skip targets only), so we take two passes to get that information.
+    // We also need to translate the thread and cpu id values into 0-based ordinals.
+    std::unordered_map<memref_tid_t, input_ordinal_t> tid2input;
+    for (int i = 0; i < static_cast<input_ordinal_t>(inputs_.size()); ++i) {
+        tid2input[inputs_[i].tid] = i;
+    }
+    std::unordered_map<uint64_t, output_ordinal_t> cpu2output;
+    std::vector<std::map<uint64_t, uint64_t>> start2stop(inputs_.size());
+    // We number the outputs according to their order in the file.
+    // XXX: Should we support some direction from the user on this?
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.cpp:711]: You mean the core_id -> output_id mapping? Does that  matter? I assumed no as the ids are an internal detail. If it does, does it also matter which thread_id is mapped to which input_id?

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 2]
```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
...
+                }
+            }
+            cur_cpu = entry.cpu;
+            cpu2output[entry.cpu] = cur_output;
+        }
+        input_ordinal_t input = tid2input[entry.thread];
+        // We'll fill in the stop ordinal in our second pass below.
+        uint64_t start = entry.instr_count;
+        uint64_t timestamp = entry.timestamp;
+        // Some entries have no instructions.
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.cpp:735]: Add why.

Side question: does it affect size of the schedule in any significant way?

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 3]
```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
...
+        if (all_sched[cur_output].size() > 0 &&
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.cpp:736]: Use !_.empty()

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 4]
```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
...
+            VPRINT(this, 3,
+                   "Output #%d: as-read segment #%zu has no instructions: skipping\n",
+                   cur_output, all_sched[cur_output].size() - 1);
+            continue;
+        }
+        all_sched[cur_output].emplace_back(schedule_record_t::DEFAULT, input, start, 0,
+                                           timestamp);
+        start2stop[input][entry.instr_count] = entry.instr_count;
+    }
+    for (int i = 0; i < static_cast<output_ordinal_t>(outputs_.size()); ++i) {
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.cpp:748]: Would it help memory usage efficiency if we collapsed each output as soon as it is created, instead of reading all outputs first? Or does that make code more complex?

Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.cpp:748]: I find it useful to use more verbose iterator vars when there are multiple iterators or access to a > 1-D data structure. Optionally use output_idx here.

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 5]
```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
...
+            input == all_sched[cur_output].back().key.input &&
+            start == all_sched[cur_output].back().start_instruction) {
+            VPRINT(this, 3,
+                   "Output #%d: as-read segment #%zu has no instructions: skipping\n",
+                   cur_output, all_sched[cur_output].size() - 1);
+            continue;
+        }
+        all_sched[cur_output].emplace_back(schedule_record_t::DEFAULT, input, start, 0,
+                                           timestamp);
+        start2stop[input][entry.instr_count] = entry.instr_count;
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.cpp:746]: Perhaps a `set` is what we need?

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 6]
```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
...
+        }
+        input_ordinal_t input = tid2input[entry.thread];
+        // We'll fill in the stop ordinal in our second pass below.
+        uint64_t start = entry.instr_count;
+        uint64_t timestamp = entry.timestamp;
+        // Some entries have no instructions.
+        if (all_sched[cur_output].size() > 0 &&
+            input == all_sched[cur_output].back().key.input &&
+            start == all_sched[cur_output].back().start_instruction) {
+            VPRINT(this, 3,
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.cpp:739]: Would the timestamp field be different for such cases?

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 7]
```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
...
+                    start_consec >= 0 ? all_sched[i][start_consec] : all_sched[i][j];
+                outputs_[i].record.emplace_back(
+                    static_cast<typename schedule_record_t::record_type_t>(toadd.type),
+                    +toadd.key.input, +toadd.start_instruction,
+                    +all_sched[i][j].stop_instruction, +toadd.timestamp);
+                start_consec = -1;
+                VDO(this, 3, {
+                    auto &added = outputs_[i].record.back();
+                    VPRINT(this, 3,
+                           "segment #%zu: input=%d start=%" PRId64 " stop=%" PRId64
```
Comment (abhinav92003) [inline clients/drcachesim/scheduler/scheduler.cpp:784]: Maybe add output-idx to log too?

Also, if one want sto see the final entries but not the as-read entries, use a higher verbosity level above or a lower here.

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 8]
```diff
@@ -1468,6 +1467,145 @@ test_replay_skip()
...
+        inputs[i].push_back(make_exit(tid));
+    }
+
+    // Synthesize a cpu-schedule file.
+    std::string cpu_fname = "tmp_test_cpu_as_traced.zip";
+    static const char *const CORE0_SCHED_STRING = "AAACCCAAACCCBBBDDD";
+    static const char *const CORE1_SCHED_STRING = "BBBDDDBBBDDDAAACCC";
+    {
+        std::vector<schedule_entry_t> sched0;
+        sched0.emplace_back(TID_BASE, 101, CPU0, 0);
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1498]: The `0` is the start instr ordinal.  You have to look at the subsequent record's start to find the count.  The 0 equals 1 (our ordinals are 1-based) so the 0..4 is 3 instrs; 4..7 and 7..infinity (9 total) are also 3.

[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1498]: Ah that's not the count of instructions executed. The "instr_count" identifier is confusing. It should be "instr_ordinal" or something to denote the point in the thread's instr stream..

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 9]
```diff
@@ -1468,6 +1467,145 @@ test_replay_skip()
...
+#if (defined(X86_64) || defined(ARM_64)) && defined(HAS_ZIP)
+    std::cerr << "\n----------------\nTesting replay as-traced from a file\n";
+    std::string path = std::string(testdir) + "/drmemtrace.threadsig.x64.tracedir";
+    std::string cpu_file =
+        std::string(testdir) + "/drmemtrace.threadsig.x64.tracedir/cpu_schedule.bin.zip";
+    // This checked-in trace has 8 threads on 9 cores, with one core starting
+    // out in a wait state.
+    static const char *const SCHED_STRING =
+        "Core #0: 3018553 \nCore #1: 3018556 \nCore #2: 3018557 \nCore #3: 3018558 "
+        "\nCore #4: 3018520 \nCore #5: 3018520 \nCore #6: 3018554 \nCore #7: 3018555 "
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1569]: I was also surprised it didn't have more.  There is a tradeoff here, though, where we want our GA CI to not take forever to run and we don't want to check in large binary files into our repository so we mostly run small apps.  This is the "threadsig" app we use for drmemtrace_samples.  This already stressed several corner cases (core 5 having to wait for core 4 right up front; merging duplicate entries) so it is a pretty good test; plus, the unit test has migration to test that aspect.  I could try to trace the threadsig app a few more times and hope to get more migration but it may not happen on this machine (plus it is work to then update the other tests that use this checked-in trace).  We also have a bigger internal proprietary test which does have a ton of migration.  I'm leaning toward just adding a comment here acknowledge the few migrations but pointed at the synthetic test having more migration.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 10]
```diff
@@ -1468,6 +1467,145 @@ test_replay_skip()
...
+        inputs[i].push_back(make_thread(tid));
+        inputs[i].push_back(make_pid(1));
+        for (int j = 0; j < NUM_INSTRS; j++)
+            inputs[i].push_back(make_instr(42 + j * 4));
+        inputs[i].push_back(make_exit(tid));
+    }
+
+    // Synthesize a cpu-schedule file.
+    std::string cpu_fname = "tmp_test_cpu_as_traced.zip";
+    static const char *const CORE0_SCHED_STRING = "AAACCCAAACCCBBBDDD";
```
Comment (abhinav92003) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:1494]: Do we verify in any way if the BBB in core0 are the instrs 7-9 in B's instrs stream? Could they be instrs 4-6?

Comment (abhinav92003) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1494]: Oh it's executes lockstep, so the fact that the index of Bs in core0 is > index of Bs in core1, that ensures it.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1569]: Ack. That's fine.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:711]: Simulation may want to preserve the NUMA relationships.  E.g., for a 2-socket chip, keep the simulated cores separated into their respective sockets to match how it was traced.  So the simulator will want to know which output streams are from which original cpuid.  We do have a proposal to add a cpuid query to the memtrace_stream_t interface for convenience but some of the output stream to core mapping is probably best done up front in the simulator before reading any actual stream records.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1498]: Yes it confused me as well reading it last week even though I wrote it (months ago).  I just added a comment at least to that field.

[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1498]: OK I also renamed the field to `start_instruction`: just a couple of uses; not likely external users are looking at this.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:711]: I don't think the input mappings matter; just the output.

Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:711]: I upgraded the XXX to a TODO: i#5843 and elaborated in the comment.

File: clients/drcachesim/scheduler/scheduler.cpp [hunk 11]
```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
...
+    // We do not have a subfile listing feature in archive_istream_t, but we can
+    // read sequentially as each record has a cpu field.
+    // This schedule_entry_t format doesn't have the stop instruction ordinal (as it was
+    // designed for skip targets only), so we take two passes to get that information.
+    // We also need to translate the thread and cpu id values into 0-based ordinals.
+    std::unordered_map<memref_tid_t, input_ordinal_t> tid2input;
+    for (int i = 0; i < static_cast<input_ordinal_t>(inputs_.size()); ++i) {
+        tid2input[inputs_[i].tid] = i;
+    }
+    std::unordered_map<uint64_t, output_ordinal_t> cpu2output;
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:708]: Not needed for this impl; removed.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:736]: Done.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:735]: The cpu_schedule file records every tracer timestamp location even if consecutive ones are on the same cpuid (it is trying to also be a record of all timestamps).  The examples I saw were things like a signal arriving right up front at thread attach before any instrs were executed so we had 2 timestamps w/ no instrs in between.  Added this to the comment.

Re: size: it affects the size of the cpu_schedule; it does not affect the scheduler internal format.  It is rare.  The bigger source if size inflation in the cpu_schedule is just repeating consecutive timestamps w/ the same cpuid.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 4 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:748]: It is actually impossible because each entry only stores its start and not its stop so we don't know if there was a migration in between two consecutive starts on one cpu.  The original use case for this cpu_schedule file was to make per-core skipping faster.  Maybe we could make a v2 that adds the stop; for now I would say let's live with these 2 passes and having to pull the whole every-timestamp entry into memory and only if we hit issues should we spend more time on it.  I augmented the comment toward the top of the file regarding this.

Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:748]: Done and s/j/shed_idx/

Comment (abhinav92003) [reply clients/drcachesim/scheduler/scheduler.cpp:748]: I don't mean that we should remove the two passes. Currently, we first read all entries for *all* outputs. I was suggesting reading for one output and then collapsing it, then reading the entries for the next output.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 5 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:746]: Yes, we just need a sorted container where we can ++ to find the next element.  Changed.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 4 above)
Comment (abhinav92003) [reply clients/drcachesim/scheduler/scheduler.cpp:748]: I guess it won't make that big of a difference since the no-instruction entries are rare.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 7 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:784]: Each output is identified by the "Read %zu..." above.  Raised verbosity above.

File: clients/drcachesim/scheduler/scheduler.cpp (hunk 4 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.cpp:748]: The no-instruction entries are already collapsed as they're read so they do not occupy memory.  The same-cpuid-consecutive entries are what are impossible to collapse until all other outputs are read.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 10 above)
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:1494]: Right.
